modules/lang-painless/src/main/resources/org/elasticsearch/painless/java.util.regex.txt

```diff
@@ -29,5 +29,31 @@ class Pattern -> java.util.regex.Pattern extends Object {
 }
 
 class Matcher -> java.util.regex.Matcher extends Object {
+  int end()
+  int end(int)
+  boolean find()
+  boolean find(int)
+  String group()
+  String group(int)
+  # String group(String)  How do I get this working?
```

Comment:
its not breaking anything: we are experimental and can do things correctly. You can have one call per arity, and it means there is no ambiguity of what method is being called. We have whitelisted over 10 of the most common java packages and this "java overloading" only impacts a rare few, which typically have another way to do it that is not ambiguous. 

For a dynamic language this overloading is a big no-no, you have no idea what method is being called!  What if you did `new StringBuilder(doc[field].fieldA).append(doc[field].fieldB)`, and then you have a mappings bug in one index or something, such that fieldA is an int? Say with a large value.

Furthermore, what java does is extremely complicated. Just look at this: https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.12.2.1 Be sure to reply back before you have torn out your eyes.

For us its just an `int`

This also means the arity of the callsite is fixed (e.g. String.xxx/3). This is important for performance, it means for method calls we only need one `guardWithTest` on the receiver, versus a shit ton (and then possibility of inline caches degrading). It also is the only reason practically that method references and lambdas can work for us with a def type, without doing something slow.

The whitelisting is done carefully so it wont be a problem. For example just whitelisting `String.valueOf(Object)` and knowing that will work fine. 

Sorry, i dont see this as something practical to change. I know that groovy does a lot of ambiguous, slow things, you probably have no clue what method is being called in many of these situations or how slow that is. That does not mean we need to make bad tradeoffs.